### PR TITLE
Strips whitespace from collection form and validates collection druid.

### LINKED ISSUE #309: Add data input validation to collection configuration form
We need to improve validation of the data input into the collection configuration form. The values entered into this form are passed onto the wasapi downloader to fetch crawl data from Archive-IT. When the values are not valid, the retrieval fails. This happens mostly in the background and it can be difficult to track down and correct the errors.

Issues we have seen:

- leading and/or trailing spaces in one or more of the values (throws off validation of druid format, collection ids)
- incorrect or non-existent collection ids

### PR DESCRIPTION
**Pull request by justinlittman**
closes #309

## Why was this change made? 🤔
More resilient to user error.

## How was this change tested? 🤨

⚡ ⚠ If this change involves consuming from other services or writing to shared file systems, test that web archive seed and crawl accessioning (and maybe even SWAP system?) works properly in [stage|qa] environment, in addition to specs. ⚡

Unit, qa

### CODE REVIEW

File: app/views/collections/_form.html.erb [hunk 1]
```diff
@@ -1,4 +1,4 @@
-<%= simple_form_for @collection do |form| %>
+<%= simple_form_for @collection, data: { turbo: false } do |form| %>
```
Comment (justinlittman) [inline app/views/collections/_form.html.erb:1]: To fix rendering of errors.

**Comment by justinlittman** [comment]
image

File: app/models/collection.rb [hunk 2]
```diff
@@ -5,12 +5,15 @@ class Collection < ApplicationRecord
...
+  # Removes extra spaces
+  auto_strip_attributes :title, :druid, :wasapi_collection_id
+
   validates_presence_of :title, :druid, :embargo_months, :fetch_start_month, :wasapi_provider_account,
                         :wasapi_collection_id
   validates :active, inclusion: { in: [true, false] }
   validates :admin_policy, inclusion: { in: %w[druid:wr005wn5739 druid:yf700yh0557] }
-  validates :druid, format: { with: /druid:.+/,
-                              message: 'must begin with druid:' }
+  validates :druid, format: { with: /\Adruid:[b-df-hjkmnp-tv-z]{2}[0-9]{3}[b-df-hjkmnp-tv-z]{2}[0-9]{4}\z/,
```
Comment (edsu) [inline app/models/collection.rb:15]: I had no idea that druids could not contain vowels!? Do `\A` and and `\z` work like `^` and `$`?

File: app/views/collections/_form.html.erb (hunk 1 above)
Comment (edsu) [reply app/views/collections/_form.html.erb:1]: What was the problem that turbo was causing here?

Comment (jcoyne) [reply app/views/collections/_form.html.erb:1]: What kind of errors?

**Review by edsu — APPROVED** [review]
Was this part of the ticket addressed in this PR, or was that decided to be out of scope for now?

> make sure that the collection druid that is entered exists and it is a collection

File: app/models/collection.rb (hunk 2 above)
Comment (jcoyne) [reply app/models/collection.rb:15]: no, `^` and `$` only match up to a new-line character. `\A` and `\z` match the whole string.

File: app/validators/collection_druid_validator.rb [hunk 3]
```diff
@@ -0,0 +1,21 @@
+# frozen_string_literal: true
+
+# Validates that collection exists.
+class CollectionDruidValidator < ActiveModel::EachValidator
+  def validate_each(record, attribute, value)
+    return unless Settings.validate_collection_druid
+    return if record.errors[attribute].present?
```
Comment (mjgiarlo) [inline app/validators/collection_druid_validator.rb:7]: If there are already errors for this attr, don't add a collection druid error to the list?

File: app/validators/collection_druid_validator.rb [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+    cocina_obj = dor_services_client.object(value).find
+    record.errors.add attribute, 'not a collection' unless cocina_obj.type == Cocina::Models::ObjectType.collection
```
Comment (mjgiarlo) [inline app/validators/collection_druid_validator.rb:10]: Can we use `... unless cocina_obj.collection?` (from the `Cocina::Models::Checkable` module) or does that not cover cases we think we might see here?

File: spec/models/collection_spec.rb [hunk 5]
```diff
@@ -58,4 +58,52 @@
...
     end
   end
+
+  context 'when validating druid' do
+    # let(:client) { instance_double(Dor::Services::Client, object: object_client) }
+    let(:object_client) { instance_double(Dor::Services::Client::Object) }
+
+    before do
+      allow(Settings).to receive(:validate_collection_druid).and_return(true)
+      # allow(Dor::Services::Client).to receive(:configure).and_return(client)
```
Comment (mjgiarlo) [inline spec/models/collection_spec.rb:68]: Cruft on lines 63 and 68 or is that worth keeping around?

File: app/validators/collection_druid_validator.rb (hunk 4 above)
Comment (edsu) [reply app/validators/collection_druid_validator.rb:10]: That helped me see where the existence of the collection was being checked -- thank you Mike.

File: app/views/collections/_form.html.erb (hunk 1 above)
Comment (justinlittman) [reply app/views/collections/_form.html.erb:1]: Turbo frames aren't added to the page. Thus, when validation failed and a page displaying the errors was being returned, that page wasn't being displayed. Hence, the errors weren't rendered to the user.

This tells turbo to stand down.

File: app/validators/collection_druid_validator.rb (hunk 3 above)
Comment (justinlittman) [reply app/validators/collection_druid_validator.rb:7]: Yes, though the motivation is to stop an invalid druid from causing an OpenAPI error when calling find. I added a comment to reflect this.

File: app/validators/collection_druid_validator.rb (hunk 4 above)
Comment (justinlittman) [reply app/validators/collection_druid_validator.rb:10]: Thanks for the suggestion. Changing.

File: spec/models/collection_spec.rb (hunk 5 above)
Comment (justinlittman) [reply spec/models/collection_spec.rb:68]: Cruft. Removing.

File: app/views/collections/_form.html.erb (hunk 1 above)
Comment (jcoyne) [reply app/views/collections/_form.html.erb:1]: What do you mean by "Turbo frames aren't added to the page."? How would they not be?

Comment (justinlittman) [reply app/views/collections/_form.html.erb:1]: https://github.com/sul-dlss/was-registrar-app/pull/464

**Pull request merged**